feat!: return Option from to_address_point and point_from_x_coord (#8…

wildjos opened the pull request:
# Motivation
Some AztecAddresses and x coordinates may not correspond to valid points on the curve.  Previously, `to_address_point` `point_from_x_coord` and `point_from_x_coord_and_sign` assumed the x coordinate were always valid, which could lead to runtime errors or potential security issues.

Now, both functions return an `Option`, returning `None` if the x coordinate does not correspond to a valid point. This makes handling invalid addresses explicit and safe.

# Key points
- `sqrt` moved from `aztec-nr/aztec/src/utils/field.nr` to `noir-protocol-circuits`
- Deleted redundant `field.nr` from `aztec-nr`
- `AztecAddress::to_address_point` now returns `Option<AddressPoint>`
- `point_from_x_coord` now returns `Option<Point>`
- `point_from_x_coord_and_sign` now returns `Option<Point>` as well
- Existing internal call sites temporarily use `.unwrap()` to maintain behaviour
- Added tests in `point.nr`, `field.nr`, and `aztec_address.nr` to verify Optionals in all relevant functions:
  - Valid x (sqrt exists)
  - Invalid x (sqrt fails, returns `None`)

# Breaking change
Any code calling `to_address_point`, `point_from_x_coord`, or `point_from_x_coord_and_sign` must now handle `Option` instead of assuming a valid point.

# Addresses
Closes #8970

nventuro commented on noir-projects/aztec-nr/aztec/src/keys/ecdh_shared_secret.nr:
```diff
@@ -24,7 +24,8 @@ pub fn derive_ecdh_shared_secret_using_aztec_address(
 ) -> Point {
-    derive_ecdh_shared_secret(ephemeral_secret, recipient_address.to_address_point().inner)
+    let addr_point = recipient_address.to_address_point().unwrap();
+    derive_ecdh_shared_secret(ephemeral_secret, addr_point.inner)
```
```suggestion
) -> Option<Point> {
    recipient_address.to_address_point().map(|addr_point |
      derive_ecdh_shared_secret(ephemeral_secret, addr_point.inner
    )
```

Let's bubble-up the Option here, since this function does not have the context to determine how to handle such an error. Its caller (`AES128::encrypt_log`) will.

wildjos replied:
I made the change to return Option<Point> as above.

nventuro commented on noir-projects/aztec-nr/aztec/src/messages/encryption/aes128.nr:
```diff
@@ -338,7 +338,7 @@ impl LogEncryption for AES128 {
 
         // Derive shared secret
-        let ciphertext_shared_secret = get_shared_secret(recipient, eph_pk);
+        let ciphertext_shared_secret = get_shared_secret(recipient, eph_pk.unwrap());
```
```suggestion
        // TODO(#8970): handle invalid ephemeral keys when decrypting to prevent DoS vectors
        let ciphertext_shared_secret = get_shared_secret(recipient, eph_pk.unwrap());
```

nventuro replied:
When leaving work that we know needs to be followed-up on, we typically create an issue explaining the problem and then link to it from the source code this way. We find this both signals clearly to the reader that there's a known problem, and provides more context to people addressing the issue.

wildjos replied:
Ok - I've created a new issue for this, see: #17158

nventuro commented on noir-projects/aztec-nr/aztec/src/utils/point.nr:
```diff
@@ -39,33 +38,33 @@ pub fn get_sign_of_point(p: Point) -> bool {
 }
 
-pub fn point_from_x_coord(x: Field) -> Point {
+pub fn point_from_x_coord(x: Field) -> Option<Point> {
```
```suggestion
/// Returns a `Point` in the Grumpkin curve given its x coordinate.
///
/// Because not all values in the field are valid x coordinates of points in the curve (i.e. there
/// is no corresponding y value in the field that satisfies the curve equation), it may not be
/// possible to reconstruct a `Point`. `Option::none()` is returned in such cases.
pub fn point_from_x_coord(x: Field) -> Option<Point> {
```

nventuro replied:
Just adding some simple docs given that we're adjusting these and that their signature is now a bit more cryptic.

nventuro commented on noir-projects/aztec-nr/aztec/src/utils/point.nr:
```diff
@@ -39,33 +38,33 @@ pub fn get_sign_of_point(p: Point) -> bool {
 /// The y coordinate can be derived from the x coordinate and the "sign" flag by solving the grumpkin curve
 /// equation for y.
 /// @param x - The x coordinate of the point
 /// @param sign - The "sign" of the y coordinate - determines whether y <= (Fr.MODULUS - 1) / 2
```
```suggestion
/// Returns a `Point` in the Grumpkin curve given its x coordinate and sign for the y coordinate.
///
/// Because not all values in the field are valid x coordinates of points in the curve (i.e. there
/// is no corresponding y value in the field that satisfies the curve equation), it may not be
/// possible to reconstruct a `Point`. `Option::none()` is returned in such cases.
///
/// @param x - The x coordinate of the point
/// @param sign - The "sign" of the y coordinate - determines whether y <= (Fr.MODULUS - 1) / 2
```

nventuro commented on noir-projects/noir-protocol-circuits/crates/types/src/address/aztec_address.nr:
```diff
@@ -52,24 +52,31 @@ impl AztecAddress {
     }
 
-    pub fn to_address_point(self) -> AddressPoint {
+    pub fn to_address_point(self) -> Option<AddressPoint> {
```
```suggestion
    /// Returns an address's `AddressPoint`, which can be used to create shared secrets with the owner
    /// of the address. If the address is invalid (i.e. it is not a properly derived Aztec address), then this
    /// returns `Option::none()`, and no shared secrets can be created.
    pub fn to_address_point(self) -> Option<AddressPoint> {
```

nventuro commented on noir-projects/noir-protocol-circuits/crates/types/src/address/aztec_address.nr:
```diff
@@ -52,24 +52,31 @@ impl AztecAddress {
-        // note: The field modulus is MAX_FIELD_VALUE + 1
-        if (!(y.lt(MAX_FIELD_VALUE / 2) | y.eq(MAX_FIELD_VALUE / 2))) {
-            y = (MAX_FIELD_VALUE + 1) - y;
+        // sqrt returns Option<Field> - return none if the sqrt fails
```
```suggestion
        // An invalid AztecAddress is one for which no y coordinate satisfies the curve equation, which we'll
        // identify by proving that the square root of y_squared does not exist.
        let mut y_opt = sqrt(y_squared);
        if y_opt.is_none() {
```

Just some small clarification on this bit - note that sqrt doesn't 'fail', but rather it proves that there's no square root in the field.

nventuro commented on noir-projects/noir-protocol-circuits/crates/types/src/utils/field.nr:
```diff
@@ -79,12 +81,37 @@ pub(crate) fn pow(x: Field, y: Field) -> Field {
 
+// Returns (false, 0) if there is no square root.
+// Returns (true, sqrt) if there is a square root.
+pub fn sqrt(x: Field) -> Option<Field> {
```
```suggestion
/// Returns Option::some(sqrt) if there is a square root, and Option::none() if there isn't.
pub fn sqrt(x: Field) -> Option<Field> {
```

I know you were just copying over the old comment, it seems to have been oudated. I also updated the comments to be a docstring by doing triple forward slashes, which will cause the LSP to display this text on hover.

nventuro commented on noir-projects/noir-protocol-circuits/crates/types/src/utils/field.nr:
```diff
@@ -1,3 +1,5 @@
+pub global KNOWN_NON_RESIDUE: Field = 5; // This is a non-residue in Noir's native Field.
```
Does this need to be `pub`?

wildjos replied:
Updated!

nventuro reviewed (approved):
Lovely, thanks a lot! Added some minor suggestions to update some of the pre-existing comments. Feel free to merge once you think this is ready.

wildjos committed:
feat!: return Option from to_address_point and point_from_x_coord (#8…

# Motivation
Some AztecAddresses and x coordinates may not correspond to valid points on the curve.  Previously, `to_address_point` `point_from_x_coord` and `point_from_x_coord_and_sign` assumed the x coordinate were always valid, which could lead to runtime errors or potential security issues.

Now, both functions return an `Option`, returning `None` if the x coordinate does not correspond to a valid point. This makes handling invalid addresses explicit and safe.

# Key points
- `sqrt` moved from `aztec-nr/aztec/src/utils/field.nr` to `noir-protocol-circuits`
- Deleted redundant `field.nr` from `aztec-nr`
- `AztecAddress::to_address_point` now returns `Option<AddressPoint>`
- `point_from_x_coord` now returns `Option<Point>`
- `point_from_x_coord_and_sign` now returns `Option<Point>` as well
- Existing internal call sites temporarily use `.unwrap()` to maintain behaviour
- Added tests in `point.nr`, `field.nr`, and `aztec_address.nr` to verify Optionals in all relevant functions:
  - Valid x (sqrt exists)
  - Invalid x (sqrt fails, returns `None`)

# Breaking change
Any code calling `to_address_point`, `point_from_x_coord`, or `point_from_x_coord_and_sign` must now handle `Option` instead of assuming a valid point.

# Addresses
Closes #8970

benesjan commented:
I see that we still have the issue [linked in TS](https://github.com/AztecProtocol/aztec-packages/blob/317f29f05c32c54ade1e0c9ac3c0d42ebec78479/yarn-project/stdlib/src/keys/derivation.ts#L128). I see that [there](https://github.com/AztecProtocol/aztec-packages/blob/317f29f05c32c54ade1e0c9ac3c0d42ebec78479/yarn-project/stdlib/src/keys/derivation.ts#L128) we throw the NotOnCurve error if the address is invalid. I guess that behavior is fine and we should just remove the TODO from there, right?
